ecommerce/products/views.py

```diff
@@ -16,3 +16,32 @@ def get_context_data(self, **kwargs):
         return context
 
 
+class ProductByTypeView(TemplateView):
+
+    template_name = 'product_by_type.html'
+
+    def get_context_data(self, product_type=None):
+        if product_type not in [c[0] for c in ProductType.CHOICES]:
+            raise Http404()
+        context = super().get_context_data()
+        context['product_type'] = (
+            models.Product.objects.filter(product_type=product_type)
+            .order_by('-created_at')
+        )
+        return context
+
+class ProductDetailView(TemplateView):
+
+    template_name = 'product_detail_view.html'
+
+    def get_context_data(self, id=None):
+        context = super().get_context_data()
+        try:
+            context['product_detail'] = (
+            models.Product.objects.filter(id=id)
+        )
+        except Product.DoesNotExist:
+            raise Http404()
```

Comment:
So  the reason this isn't throwing a 404 error is because you're using `filter`. If there are no matching products for filter, it will return an empty queryset. If you use `get`, it throws and exception if there is no matching object. It also throws an exception if there are multiple matching objects but that won't be a problem in this case cos you're accessing it by id which is unique. Have a read through this section: https://docs.djangoproject.com/en/3.1/topics/db/queries/#retrieving-objects

Also, django actually has a nice shortcut method which encapsulates the `try/catch`   logic you've used here. This would tidy up the logic in this view nicely https://docs.djangoproject.com/en/3.1/topics/http/shortcuts/#get-object-or-404